### ISSUE TITLE
Post-deployment operations are complex

### ISSUE THREAD

**Issue by abonillasuse** [SUSE/DeepSea#236]
Replacing a drive after deploying with DeepSea means:
- Clean up
Make sure the OSD is not running/mounted
	systemctl stop ceph-osd@5.service
	umount /dev/sdb1
	ceph osd crush reweight osd.5 0
	ceph osd crush remove osd.5 && ceph auth del osd.5n && ceph osd rm osd.5

- Deploy new drive
Ensure the (replacement) drive is visible, i.e. /dev/sdb
Clear any partitions from it
	parted -s /dev/sdb mklabel gpt
	partprobe
Modify the DeepSea proposal somehow. (HOW!?)
Run stage.1 to stage.3

I would appreciate a:
salt-run state.orch ceph.operation.removeosd 5
- - which runs all the cleanup from above
salt-run state.orch ceph.operation.pre-rescan
- - which moves whatever .yml to a bkp location
- - provides a new .yml showing the new disk

run stage 2 to 3

**Comment by swiftgist** [comment]
The short requirement is you want the single OSD removal utility.  It looks like this will be a side effect of implementing bluestore/dmcrypt.  

The main dilemma is removing separate journals (or wals and dbs) from devices other than the OSD drive when the OSD has completely failed.  Currently, DeepSea has no way to know which journal on an SSD/NVMe is the right one to delete.  The current direction with the bluestore/dmcrypt is moving the complexity from

/srv/salt/ceph/osd/*.sls

into the osd salt module.  This will include a fourth, unnamed step to collect and store the partition names on creation.  This procedure would look like

partition
prepare
activate
TBD

The favored location  for storing this information looks like custom grains.

As far as the utility, it may be a runner rather than an orchestration.  I can imagine 

```
salt-run remove.osd 11
```
unless that seems to similar to ceph osd crush remove command.

**Comment by abonillasuse** [comment]
@swiftgist this would work. Additionally, a replacement OSD means also a new discovery and addition of disks. How can we implement that without  overwriting other proposals, etc

**Comment by swiftgist** [comment]
That would be the task of the utility.  This would include "fixing" the configuration in the Salt pillar by removing the absent devices.  

For paranoia's sake and giving you a "Get out of Jail" card (Monopoly reference for those that don't know), I am open to hearing how you want what where.  That is, name/location of the previous file.  (Should we ever start using git for configuration management, that would be moot.)  I expect to keep the previous copy.

Generally, the discovery portion that writes these files is consistent.  However, if manual editing has been done (or the configuration is pulled in from a ceph-deploy cluster), then those changes need to be preserved.

**Comment by swiftgist** [comment]
Looks like I'm heading in the direction of 

```
salt-run osd.remove 11
```
I need this functionality for Stage 5 to work with dmcrypt.  The side effect will be creating this runner.  I haven't pushed the branch yet.

**Comment by jschmid1** [comment]
That should be resolved with the latest changes to osd.py @swiftgist ?

**Comment by swiftgist** [comment]
The remove runner and module addresses the removal parts.  The proposal runner should address the profile management with new devices.  The migrate orchestration addresses reconfiguration.  All are available in 0.7.13.

Quick list of commands:

```
salt-run remove.osd ID
salt 'node*' osd.remove ID
salt-run proposal.help
salt-run state.orch ceph.migrate.osds
```
